tests/wasm/coverage/annotate.ts

```diff
@@ -0,0 +1,231 @@
+import {
+  Parser,
+  Statement,
+  NodeKind,
+  Node,
+  Source,
+  FunctionDeclaration,
+  BlockStatement,
+  IfStatement,
+  ForStatement,
+  ForOfStatement,
+  WhileStatement,
+  DoStatement,
+  SwitchStatement,
+  SwitchCase,
+} from 'assemblyscript'
+
+import { InstrumentedLines } from './'
+
+/**
+ * Annotates a given assembly script source node with a `trace` instructions before each statement.
+ * The trace instruction contains the filename as well as the covered line numbers
+ * `trace($file, $from_line, $to_line)
+ *
+ * @param source A source node e.g. from parsing an assembly script file
+ * @returns a bitmap of instrumented lines in the file
+ */
+export function annotateSource(source: Source): InstrumentedLines {
+  const line_count = source.text.split('\n').length
+  const instrumentedLines: InstrumentedLines = new Array(line_count).fill(false)
+  annotateNodeRecursively(source, source.text, instrumentedLines)
+  return instrumentedLines
+}
+
+// Returns a parsed `trace` expression that can be inserted in the AST to annotate the given statement
+function traceExpressionFor(statement: Statement, file_content: string, instrumentedLines: InstrumentedLines) {
+  const range = statement.range
+  const start = lineNumberForCharacterIndex(file_content, range.start)
+  // Some multiline statements (e.g. function declarations) shouldn't be covered as soon as the first line is evaluated
+  const end = isRangeFromStartToEnd(statement) ? lineNumberForCharacterIndex(file_content, range.end) : start
+  for (let i = start; i <= end; i++) {
+    instrumentedLines[i] = true
+  }
+
+  // Use a parser to create the `trace` expression from a string
+  const trace = `trace('${range.source.normalizedPath}', ${start}, ${end})`
+  const parser = new Parser()
+  parser.parseFile(trace, 'trace.ts', true)
+  return parser.sources[0].statements[0]
+}
+
+// Depending on the Node we might have to annotate nested notes recursively
+function annotateNodeRecursively(node: Node, content: string, instrumentedLines: InstrumentedLines) {
+  switch (node.kind) {
+    case NodeKind.SWITCHCASE:
+    case NodeKind.BLOCK:
+    case NodeKind.SOURCE: {
+      const castedNode = node as Source | BlockStatement | SwitchCase
+      castedNode.statements = castedNode.statements.flatMap((s) => {
+        annotateNodeRecursively(s, content, instrumentedLines)
+        return [traceExpressionFor(s, content, instrumentedLines), s]
+      })
+      break
+    }
+    case NodeKind.FUNCTIONDECLARATION: {
+      const fn = node as FunctionDeclaration
+      fn.body ? annotateNodeRecursively(fn.body, content, instrumentedLines) : {}
```

Comment:
Interesting way to write an `if` statement. I guess TS doesn't like `fn.body && annotateNodeRecursively(...)`.